### ISSUE TITLE
HttpClient InputStream.read() hangs intermittently before end of response

### ISSUE THREAD

**Issue by dennis-yemelyanov** [eclipse/jetty.project#11563]
**Description**
For some websites reading from `Source.asInputStream(contentSource)` stop after a certain point (`InputStream.read()` blocks indefinitely). It happens intermittently (not every time I run the app, but most of the times) and can be "fixed" by adding `Thread.sleep(1)`. Enabling Jetty debug logging also "fixes" it probably for the same reason as adding a sleep does.

**How to reproduce?**
Run the following code, which is likely to stop reading at some point, at least for me it does after 47392 bytes:
```java
package com.example;

import java.io.InputStream;
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.Request;
import org.eclipse.jetty.client.Response.ContentSourceListener;
import org.eclipse.jetty.io.Content.Source;

public class App {
    public static void main(String[] args) throws Exception {
        HttpClient httpClient = new HttpClient();

        httpClient.start();

        Request req = httpClient.newRequest("https://bing.com").onResponseContentSource(new ContentSourceListener() {
            @Override
            public void onContentSource(org.eclipse.jetty.client.Response response, Source contentSource) {
                try {
                    // Uncomment to "fix" the issue
                    //Thread.sleep(1);

                    try (InputStream inputStream = Source.asInputStream(contentSource)) {
                        byte[] buffer = new byte[1024];

                        System.out.println("Started reading...");

                        int totalBytes = 0;

                        for (int length; (length = inputStream.read(buffer)) != -1;) {
                            totalBytes += length;

                            System.out.println("Read:\t" + length + " bytes\tTotal:\t" + totalBytes + " bytes");
                        }

                        System.out.println("!!! response length: " + totalBytes + " !!!");

                        //System.exit(0);
                    }
                } catch (Throwable e) {
                    e.printStackTrace();
                }
            }
        });

        req.send(result -> {}); // dummy handler to make the call async. Is there a better way?

        System.out.println("Press Enter to dump httpClient...");
        
        System.console().readLine();

        System.out.println(httpClient.dump());
    }
}

```
> ... previous bytes read ...
> Read:   1024 bytes      Total:  41063 bytes
> Read:   298 bytes       Total:  41361 bytes
> Read:   1024 bytes      Total:  42385 bytes
> Read:   1024 bytes      Total:  43409 bytes
> Read:   1024 bytes      Total:  44433 bytes
> Read:   1024 bytes      Total:  45457 bytes
> Read:   1024 bytes      Total:  46481 bytes
> Read:   911 bytes       Total:  47392 bytes
... no progress after this ...

After uncommenting `Thread.sleep(1)`, most likely all bytes can be read successfully:

> ... previous bytes read ...
> Read:   1024 bytes      Total:  244497 bytes
> Read:   1024 bytes      Total:  245521 bytes
> Read:   1024 bytes      Total:  246545 bytes
> Read:   1024 bytes      Total:  247569 bytes
> Read:   1024 bytes      Total:  248593 bytes
> Read:   1024 bytes      Total:  249617 bytes
> Read:   1024 bytes      Total:  250641 bytes
> Read:   449 bytes       Total:  251090 bytes
!!! response length: 251090 !!!

Since I'm not able to get a debug log for the bad run, I'm attaching the result of `HttpClient.dump()` for both good and bad runs:
[dump bad.txt](https://github.com/jetty/jetty.project/files/14737037/dump.bad.txt)
[dump good.txt](https://github.com/jetty/jetty.project/files/14737038/dump.good.txt)

Is there any potential issue in Jetty that might be causing this? Or am I using some API incorrectly?

I was also trying to step through the code, but any interference like that causes the issue to no longer repro.

**Jetty version(s)**
12.0.7

**Jetty Environment**

**Java version/vendor** `(use: java -version)`
openjdk version "17.0.7" 2023-04-18
OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7)
OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode, sharing)

**OS type/version**
Windows 11

**Comment by dennis-yemelyanov** [comment]
Another observation is that the URL in question returns a redirect. And when hitting the target URL instead of the original one, the issue doesn't happen.

Not sure if the redirect just has some impact on timing, or could there be a potential issue with how redirects are handled by Jetty?

**Comment by sbordet** [comment]
You are using the APIs incorrectly.

Please read the documentation at: https://eclipse.dev/jetty/documentation/jetty-12/programming-guide/index.html#pg-client-http-content-response.

The problem is that you are using the non-blocking APIs, but then you block inside `onContentSource()` because you are using `InputStream`.

When the client is faster than the server, `InputStream` will block and halt completely the processing of the response content because `onContentSource()` does not return.
By adding sleeps or DEBUG logging, you slow the client enough to (by chance) never block, so it is able to read the full request without blocking and returning from `onContentSource()`.

If you want to use `InputStream`, the correct solution is to use `InputStreamResponseListener` as documented.

I'll improve the documentation.

**Comment by dennis-yemelyanov** [comment]
Thank you @sbordet for the explanation!

I actually have looked at the `InputStreamResponseListener` documentation before, but was confused by the example because it seems to block on this line:
```java
// Wait for the response headers to arrive.
Response response = listener.get(5, TimeUnit.SECONDS);
```
I have these requirements:
1. Avoid blocking while waiting for the server to send headers or start sending response (since it can take a long time).
2. Avoid buffering and instead do streaming using `InputStream` (since the response can be pretty big).

If my understanding about `InputStreamResponseListener.get()` blocking the current thread until headers arrive is correct, what is the recommended API for me to use given these requirements? If this scenario can be documented, it will be very helpful!

**Comment by sbordet** [comment]
`InputStreamResponseListener` does block on `get(5, SECONDS)`, as explained in the javadocs.

You do not need to call this blocking method, however.
You can just call `listener.getInputStream()` directly, and when content is available you will read it from the `InputStream`.
The problem may be that you won't know if it is content for a 200 OK response, or for a 500 Server Error response.

You want a fully asynchronous scenario, which is documented at the same documentation link I gave above.

You can be notified of the response status code in `onResponseHeaders()`, or in `onResponseContentSource()` and there decide whether you want to read the content or discard it.

I would recommend streaming the content, but without using `InputStream`, which only offers a blocking API.

For example, to write the response content to a file via streaming I would use `Response.AsyncContentListener`, write every chunk to the file, and when the write completes demand for more response content -- all asynchronous no `InputStream` necessary.

**Comment by dennis-yemelyanov** [comment]
Thank you for providing these suggestions! Using `InputStreamResponseListener` along with `onResponseHeaders()` was my first thought as well, but when I originally tried it, the processing got stuck before even reading the first byte. Now, with your explanation, I realize that I was not too far off, but what I was missing was the reading had to be done on a new thread.

Then does the following code look like correct usage? It works, but I want to make sure it's not just luck:
```java
InputStreamResponseListener listener = new InputStreamResponseListener();

Request req = httpClient.newRequest("https://bing.com").onResponseHeaders(new HeadersListener() {
    @Override
    public void onHeaders(Response response) {
        new Thread(() -> { // key element, without a new thread inputStream.read() may deadlock
            try (InputStream inputStream = listener.getInputStream()) {
                byte[] buffer = new byte[1024];

                System.out.println("Started reading...");

                int totalBytes = 0;

                for (int length; (length = inputStream.read(buffer)) != -1;) {
                    totalBytes += length;

                    System.out.println("Read:\t" + length + " bytes\tTotal:\t" + totalBytes + " bytes");
                }

                System.out.println("!!! response length: " + totalBytes + " !!!");

                System.exit(0);
            } catch (Throwable e) {
                e.printStackTrace();
            }
        }).start();                                          
    }
});

req.send(listener);
```
One piece I'm still missing, but would like to understand more about is, what is the intended usage scenario for `Response.Source.asInputStream(contentSource)`? If `contentSource` represents non-blocking API, incompatible with blocking `InputStream`, then why might someone want to call this method at all? I guess the same question applies to other methods like `Response.Source.asString(contentSource)`.

Interestingly, executing the reading code on a separate thread inside `onContentSource` throws a Serialized invocation error (although a similar example with `Source.asString()` seems to be working fine):

```java
@Override
public void onContentSource(org.eclipse.jetty.client.Response response, Source contentSource) {
    new Thread(() -> {
        try (InputStream inputStream = Source.asInputStream(contentSource)) {
            byte[] buffer = new byte[1024];

            System.out.println("Started reading...");

            int totalBytes = 0;

            // throws Serialized invocation error
            for (int length; (length = inputStream.read(buffer)) != -1;) {
                totalBytes += length;

                System.out.println("Read:\t" + length + " bytes\tTotal:\t" + totalBytes + " bytes");
            }

            System.out.println("!!! response length: " + totalBytes + " !!!");
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }).start();
}
```

**Comment by sbordet** [comment]
@dennis-yemelyanov read the documentation, yes? 😉 

You don't need to spawn a new thread like your snippet above.
You just:

```java
InputStreamResponseListener listener = new InputStreamResponseListener();

httpClient.newRequest("http://domain.com/path")
    .send(listener);

try (InputStream inputStream = listener.getInputStream()) {
  // read here.
}
```

Blocking another thread or blocking the main thread is the same.

There are many use cases for converting a `Content.Source` to an `InputStream` for example to re-use filtering/transforming `InputStream` such as `GzipInputStream`, etc.

`Content.Source` is used on the client, but also on the server, where the threading requirements are different.

Even on the client, you may want to obtain the `Content.Source`, but not read from it yet, setup some other API that require `InputStream`, and then link the two.

The APIs are not incompatible, they are just different, and the client listeners have a threading requirement that is common in most asynchronous libraries.

**Comment by dennis-yemelyanov** [comment]
> @dennis-yemelyanov read the documentation, yes? 😉

Of course I did! And I have to say it's really good, I like how it explains general principles and concepts that users need to understand, and it goes well beyond just javadocs.

Was just having a bit of trouble applying it to my specific requirement, which is essentially "let the caller call a method without blocking them, and notify them (on whatever thread) when their parsed response is ready (where parsing is done using another library that takes in an `InputStream`)".

Now it's more clear that I was missing the requirement not to block in the event handler code, which actually makes sense the more I think about it. I guess it's possible that some of the asynchronous libraries I've used before were executing the handler code on some separate thread pool thread already, or made it harder to block by not providing blocking methods in this context. But now I know I should always confirm this requirement and make sure my code aligns with it
